core/src/enc/bases.rs

```diff
@@ -0,0 +1,201 @@
+    NullDataScript,
+}
+
+/// Impl explicitly because FromBase58CheckError doesn't implement the std error format
+impl From<FromBase58CheckError> for EncodingError {
+    fn from(e: FromBase58CheckError) -> Self {
+        EncodingError::B58Error(e)
+    }
+}
+
+/// A simple result type alias
+pub type EncodingResult<T> = Result<T, EncodingError>;
+
+/// Encode a byte vector to bech32. This function expects `v` to be a witness program, and will
+/// return an `UnknownScriptType` if it does not meet the witness program format.
+pub fn encode_bech32(hrp: &str, v: u8, h: &[u8]) -> EncodingResult<String> {
+    let mut v = vec![u5::try_from_u8(v)?];
+    v.extend(&h.to_base32());
+    b32_encode(hrp, &v).map_err(|v| v.into())
+}
+
+/// Decode a witness program from a bech32 string. Caller specifies an expected HRP. If a
+/// different HRP is found, returns `WrongHRP`.
+pub fn decode_bech32(expected_hrp: &str, s: &str) -> EncodingResult<(u8, Vec<u8>)> {
+    let (hrp, data) = b32_decode(&s)?;
+    if hrp != expected_hrp {
+        return Err(EncodingError::WrongHRP {
+            got: hrp,
+            expected: expected_hrp.to_owned(),
+        });
+    }
+
+    // Extract the witness version and payload
+    let (v, p) = data.split_at(1);
+    let payload = Vec::from_base32(&p)?;
+
+    Ok((v[0].to_u8(), payload))
+}
+
+/// Encodes a byte slice to base58check with the specified version byte.
+pub fn encode_base58(version: u8, v: &[u8]) -> String {
+    v.to_base58check(version)
+}
+
+/// Decodes base58check into a byte string. Returns a `FromBase58CheckError` if the checksum or
+/// encoding is wrong. Returns a `WrongVersion` if it decodes an unexpected version.
+pub fn decode_base58(expected_version: u8, s: &str) -> EncodingResult<Vec<u8>> {
+    let (version, data) = s.from_base58check()?;
+    if version != expected_version {
+        return Err(EncodingError::WrongVersion {
+            got: version,
+            expected: expected_version,
+        });
+    };
+    Ok(data)
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[test]
+    fn it_should_encode_and_decode_bech32() {
+        let hrp = "bc";
+        let addrs = [
+            "bc1q233q49ve8ysdsztqh9ue57m6227627j8ztscl9",
+            "bc1qaqm8wh8sr6gfx49mdpz3w70z48xdh0pzlf5kgr",
+            "bc1qjl8uwezzlech723lpnyuza0h2cdkvxvh54v3dn",
+            "bc1qn0q63kkp3rj5wyap5fzymlvat28cu2s87tgzu6",
+            "bc1qnsupj8eqya02nm8v6tmk93zslu2e2z8chlmcej",
+            "bc1qmcwrdlcqrwcfs6654m8zvmzdmtpuvcxuzn9ahy",
+            "bc1qvyyvsdcd0t9863stt7u9rf37wx443lzasg0usy",
+            "bc1qza7dfgl2q83cf68fqkkdd754qx546h4u9vd9tg",
+            "bc1qwqdg6squsna38e46795at95yu9atm8azzmyvckulcc7kytlcckxswvvzej",
+            "lnbc20m1pvjluezpp5qqqsyqcyq5rqwzqfqqqsyqcyq5rqwzqfqqqsyqcyq5rqwzqfqypqhp58yjmdan79s6qqdhdzgynm4zwqd5d7xmw5fk98klysy043l2ahrqscc6gd6ql3jrc5yzme8v4ntcewwz5cnw92tz0pc8qcuufvq7khhr8wpald05e92xw006sq94mg8v2ndf4sefvf9sygkshp5zfem29trqq2yxxz7"
```

Comment:
It works with a lightning invoice which is significantly longer than what is allowed by for Bitcoin consensus.